linera-core/src/client/mod.rs

```diff
@@ -2469,60 +2356,131 @@ where
             }
         }
         self.update_from_info(&info);
-        let manager = &info.manager;
 
-        // If there is a validated block in the current round, finalize it.
-        if let Some(certificate) = &manager.requested_locked {
-            if certificate.round == manager.current_round {
+        let executed_block = if let Some(certificate) = &info.manager.requested_locked {
+            if certificate.round == info.manager.current_round {
+                // If there is a validated block in the current round, finalize it.
                 let committee = self.local_committee().await?;
                 match self.finalize_block(&committee, *certificate.clone()).await {
                     Ok(certificate) => return Ok(ClientOutcome::Committed(Some(certificate))),
                     Err(ChainClientError::CommunicationError(_)) => {
                         // Communication errors in this case often mean that someone else already
                         // finalized the block.
-                        let timestamp = manager.round_timeout.ok_or_else(|| {
+                        let timestamp = info.manager.round_timeout.ok_or_else(|| {
                             ChainClientError::BlockProposalError(
                                 "Cannot propose in the current round.",
                             )
                         })?;
                         return Ok(ClientOutcome::WaitForTimeout(RoundTimeout {
                             timestamp,
-                            current_round: manager.current_round,
+                            current_round: info.manager.current_round,
                             next_block_height: info.next_block_height,
```

Comment:
Is it correct? if someone else finalized the block then the actual next block height is next _next_ of what we know last.

Reply:
Right, I think the comment is misleading: It _could_ be that someone else finalized it; it could also be that somehow another round started (e.g. if it's a multi-leader round, someone could have made a higher-round proposal).

`WaitForTimeout` will get interrupted by the `NewRound` or `NewBlock` notification as appropriate, so in either case we will then end up in this function again, with the updated information.

Reply:
If we queried for the latest `ChainManagerInfo`, do we risk making anything worse? I understand that it might not be necessary but is it incorrect?